pkg/factory/factory.go

```diff
@@ -177,37 +177,41 @@ func (rf *resourceFactory) GetDeviceProvider(dt types.DeviceType) types.DevicePr
 	}
 }
 
-// GetDeviceFilter unmarshal the "selector" values from ResourceConfig and returns an instance of DeviceSelector based on
-// DeviceType in the ResourceConfig
-func (rf *resourceFactory) GetDeviceFilter(rc *types.ResourceConfig) (interface{}, error) {
-	switch rc.DeviceType {
-	case types.NetDeviceType:
-		netDeviceSelector := &types.NetDeviceSelectors{}
-
-		if err := json.Unmarshal(*rc.Selectors, netDeviceSelector); err != nil {
-			return nil, fmt.Errorf("error unmarshalling NetDevice selector bytes %v", err)
+// parseObjectOrSlice unmarshal's the "Selector" values from the ResourceConfig into a slice of *DeviceSelectors.
+// Each *DeviceSelector has been converted to any before being returned. parseObjectOrSlice will parse both
+// kinds of valid "selector" values - a slice or a single object.
+func parseObjectOrSlice[O types.NetDeviceSelectors | types.AccelDeviceSelectors | types.AuxNetDeviceSelectors](
+	rc *types.ResourceConfig) ([]any, error) {
+	slice := make([]*O, 1)
```

Comment:
this should be : `make([]O, 1)` ?

else `selectors: null` is valid and it will be parsed without error.

Reply:
Thanks, I had not considered this case. I will have to see if `make([]0, 1)` is easy to convert into the expected function output, or if manually checking that `slice[0] != nil` is more straightforward. The solution should hopefully also handle a case like `selectors: [{ /* something valid */}, nil]` and so forth.

Reply:
I was attempting to write an additional unit test for this `selectors: null` case, and ran into a small confusion. By `selectors: null` is that equivalent to an end user writing a full on config akin to:
```json
{
    "resourceList": [{
            "resourceName": "intel_sriov_netdevice",
            "selectors": "null"
        }]
}
``` 
or
```json
{
    "resourceList": [{
            "resourceName": "intel_sriov_netdevice",
            "selectors": null
        }]
}
```
? For the latter case, I am not sure how to encode the `null` into a Go program. I would want to use `nil`, but I'm not sure how to embed that into a JSON string/[]byte.

In at least the first case (a string with value `"null"`), it seems almost impossible to detect this case within this method. From the [json docs](https://pkg.go.dev/encoding/json#Unmarshal):

> The JSON null value unmarshals into an interface, map, pointer, or slice by setting that Go value to nil. Because null is often used in JSON to mean “not present,” unmarshaling a JSON null into any other Go type has no effect on the value and produces no error. 

When I run this case against the head of `master`, the log line is:
```
manager.go:95] unable to get SelectorObj from selectors list:'"null"' for deviceType: netDevice error: error unmarshalling NetDevice selector bytes json: cannot unmarshal string into Go value of type types.NetDeviceSelectors
```
As the log line shows, there isn't anything special about the `"null"` string compared to any other string value.

For the latter case, the pod crashes. Using the upstream code (without this PR), the logs are:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xd58630]

goroutine 1 [running]:
github.com/k8snetworkplumbingwg/sriov-network-device-plugin/pkg/factory.(*resourceFactory).GetDeviceFilter(0x40002aa600?, 0x4000876c60)
	/usr/src/sriov-network-device-plugin/.gopath/src/github.com/k8snetworkplumbingwg/sriov-network-device-plugin/pkg/factory/factory.go:187 +0x70
main.(*resourceManager).readConfig(0x40008d02a0)
	/usr/src/sriov-network-device-plugin/.gopath/src/github.com/k8snetworkplumbingwg/sriov-network-device-plugin/cmd/sriovdp/manager.go:92 +0x244
main.main()
	/usr/src/sriov-network-device-plugin/.gopath/src/github.com/k8snetworkplumbingwg/sriov-network-device-plugin/cmd/sriovdp/main.go:45 +0xc8
```

From this, I assume the issue is that `rc.Selectors` is a null pointer, which gets dereferenced as part of
```go
json.Unmarshal(*rc.Selectors, netDeviceSelector)
```

The exact same logs occur if the config map doesn't have any `"selectors"` field at all. This can be fixed by checking `rc.Selectors` is not `nil` both here and in the `resourceManager`.

The PR has the same behavior for these cases as the `master` branch.

Reply:
I think we should only cover the nil case and not the `"nil"` one that should let the device plugin crash..

Reply:
agree, lets handle it in a different PR.
some checks will need to be added to prevent these cases.